### PR TITLE
Get working with k8s.core 2 and above

### LINKED ISSUE #33: Making the collection working with community.kubernetes 2.0.0
##### SUMMARY
Relates to https://github.com/ansible-collections/community.kubevirt/pull/25.

The collection has a dependency - the [community.kubernetes](https://github.com/ansible-collections/community.kubernetes) collection.
There were breaking changes introduced in `community.kubernetes 2.0.0` and `community.kubevirt` cannot work with it.

Currently the community.kubernetes versions are restricted in [galaxy.yml](https://github.com/ansible-collections/community.kubevirt/blob/main/galaxy.yml):
```
dependencies:
  community.kubernetes: '>=1.0.0,<2.0.0'
```

1) The issue is to make the collection working with community.kubernetes 2.0.0+

This will be a breaking change also, so will require the next major version and prior announcement some time before the release.

**OR**
2) Evaluate if the dependency is really needed? Maybe it's sensible to get rid of it? (I don't use the collections, so hard to say)

##### ISSUE TYPE
- Bug Report

### PR DESCRIPTION
**Pull request by snecklifter**
##### SUMMARY
* Drop Python 2 support
* Kubernetes Core compatibility
* Some OpenShift -> Kubernetes renaming

Fixes #33 

##### ISSUE TYPE
- Feature Pull Request

### CODE REVIEW

**Comment by snecklifter** [comment]
@Andersson007 @mnecas this is what I've managed so far. The unit tests for 2.9 are failing and some are being skipped for the rest. Would appreciate some eyes as this is obviously a large breaking change and working with python isn't my day job. But its a start... :)

**Comment by Andersson007** [comment]
@snecklifter thanks for the PR!
~~I see units against 2.9 failed~~

**Comment by Andersson007** [comment]
also this definitely requires manual testing

**Comment by Andersson007** [comment]
Several important things:
1. When the PR is ready for merge, i would announce the changes in a separate dedicated PR (let's call it `the announcement PR`) that will add only a similar changelog fragment announcing the breaking changes that will be introduced in 2.0.0, etc.
2.  We should merge the `announcement PR` ASAP and release the fragment with the next minor / patch release.
3.  We should have a porting guide as well.
4.  IMO, we should plan 2.0.0 release right after Ansible 5.0.0 (IIRC in November) because
  a) users should be informed some time before, have porting guide ready, etc.
  b) If we release 2.0.0 after Ansible 5.0.0, 2.0.0 will be included in Ansible 6.0.0 but will be available for manual installation from Galaxy.
It will allow to mitigate the move and avoid mass production failings and anger (as well as stress for us).
5. We should merge this PR right before 2.0.0 release.
6. Would be a great thing releasing an alpha version first and waiting for feedback from users (we can announce this in Bullhorn, IRC, etc.). See community.dns release history as an example. It will prevent inclusion of the release in Ansible and accident installations via `ansilbe-galaxy --latest` but will allow interested users to install and test the version through specifying that alpha version directly.

File: changelogs/fragments/k8s-core-2-compatibility.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+major_changes:
+  - Drop Python 2.7 compatibility
+  - Migrate to kubernetes.core collection
+  - Change references from OpenShift to Kubernetes
```
Comment (Andersson007) [inline changelogs/fragments/k8s-core-2-compatibility.yml]:
```suggestion
  - Drop Python 2.7 compatibility (https://github.com/ansible-collections/community.kubevirt/issues/33).
  - Migrate from the ``community.kubernetes`` to ``kubernetes.core`` collection dependency (https://github.com/ansible-collections/community.kubevirt/issues/33).
  - Change references from ``OpenShift`` to ``Kubernetes`` in the documentation (https://github.com/ansible-collections/community.kubevirt/issues/33).
```
Or something like that

**Comment by snecklifter** [comment]
> Several important things:
> 
> 1. When the PR is ready for merge, i would announce the changes in a separate dedicated PR (let's call it `the announcement PR`) that will add only a similar changelog fragment announcing the breaking changes that will be introduced in 2.0.0, etc.
> 2. We should merge the `announcement PR` ASAP and release the fragment with the next minor / patch release.
> 3. We should have a porting guide as well.
> 4. IMO, we should plan 2.0.0 release right after Ansible 5.0.0 (IIRC in November) because
>    a) users should be informed some time before, have porting guide ready, etc.
>    b) If we release 2.0.0 after Ansible 5.0.0, 2.0.0 will be included in Ansible 6.0.0 but will be available for manual installation from Galaxy.
>    It will allow to mitigate the move and avoid mass production failings and anger (as well as stress for us).
> 5. We should merge this PR right before 2.0.0 release.
> 6. Would be a great thing releasing an alpha version first and waiting for feedback from users (we can announce this in Bullhorn, IRC, etc.). See community.dns release history as an example. It will prevent inclusion of the release in Ansible and accident installations via `ansilbe-galaxy --latest` but will allow interested users to install and test the version through specifying that alpha version directly.

This all makes sense to me, would just appreciate some input on the failing tests if anyone has the bandwidth.

**Comment by Andersson007** [comment]
> This all makes sense to me, would just appreciate some input on the failing tests if anyone has the bandwidth.

OK, i'll try to take a look on Monday

File: plugins/module_utils/kubevirt.py [hunk 2]
```diff
@@ -6,16 +6,36 @@
...
+try:
+    from ansible_collections.kubernetes.core.plugins.module_utils.args_common import AUTH_ARG_SPEC, list_dict_str
+    from ansible_collections.kubernetes.core.plugins.module_utils.common import K8sAnsibleMixin, get_api_client
+    HAS_KUBERNETES_COLLECTION = True
+except ImportError as e:
+    HAS_KUBERNETES_COLLECTION = False
+    k8s_collection_import_exception = e
+    K8S_COLLECTION_ERROR = traceback.format_exc()
+    K8sAnsibleMixin = object
+    AUTH_ARG_SPEC = {}
```
Comment (tadeboro) [inline plugins/module_utils/kubevirt.py]: You do not have to protect imports from collections that are listed as requirements.

File: plugins/module_utils/kubevirt.py [hunk 3]
```diff
@@ -132,9 +152,44 @@ def __cmp__(self, other):
...
+        self.fail_json = self.module.fail_json
+        self.fail = self.module.fail_json
+        self.exit_json = self.module.exit_json
+
+        if not HAS_KUBERNETES_COLLECTION:
+            self.fail_json(
+                msg="The kubernetes.core collection must be installed",
+                exception=K8S_COLLECTION_ERROR,
+                error=to_native(k8s_collection_import_exception)
+            )
```
Comment (tadeboro) [inline plugins/module_utils/kubevirt.py]: This is not needed (see the previous comment).

File: plugins/modules/kubevirt_cdi_upload.py [hunk 4]
```diff
@@ -84,9 +84,19 @@
...
+try:
+    from ansible_collections.kubernetes.core.plugins.module_utils.common import K8sAnsibleMixin, get_api_client
+    from ansible_collections.kubernetes.core.plugins.module_utils.args_common import AUTH_ARG_SPEC
+    HAS_KUBERNETES_COLLECTION = True
+except ImportError as e:
+    HAS_KUBERNETES_COLLECTION = False
+    k8s_collection_import_exception = e
+    K8S_COLLECTION_ERROR = traceback.format_exc()
+    K8sAnsibleMixin = object
+    AUTH_ARG_SPEC = {}
```
Comment (tadeboro) [inline plugins/modules/kubevirt_cdi_upload.py]: No protection is needed here.

File: plugins/modules/kubevirt_cdi_upload.py [hunk 5]
```diff
@@ -113,9 +123,22 @@
...
+            supports_check_mode=True,
+        )
+        self.fail_json = self.module.fail_json
+
+        if not HAS_KUBERNETES_COLLECTION:
+            self.module.fail_json(
+                msg="The kubernetes.core collection must be installed",
+                exception=K8S_COLLECTION_ERROR,
+                error=to_native(k8s_collection_import_exception)
+            )
```
Comment (tadeboro) [inline plugins/modules/kubevirt_cdi_upload.py]: Not needed.

File: plugins/modules/kubevirt_preset.py [hunk 6]
```diff
@@ -87,8 +87,14 @@
...
-
-from ansible_collections.community.kubernetes.plugins.module_utils.common import AUTH_ARG_SPEC
+try:
+    from ansible_collections.kubernetes.core.plugins.module_utils.args_common import AUTH_ARG_SPEC
+    HAS_KUBERNETES_COLLECTION = True
+except ImportError as e:
+    HAS_KUBERNETES_COLLECTION = False
+    k8s_collection_import_exception = e
+    K8S_COLLECTION_ERROR = traceback.format_exc()
+    AUTH_ARG_SPEC = {}
```
Comment (tadeboro) [inline plugins/modules/kubevirt_preset.py]: No protection is needed.

File: plugins/modules/kubevirt_pvc.py [hunk 7]
```diff
@@ -250,9 +250,20 @@
...
+try:
+    from ansible_collections.kubernetes.core.plugins.module_utils.common import K8sAnsibleMixin, get_api_client
+    from ansible_collections.kubernetes.core.plugins.module_utils.args_common import AUTH_ARG_SPEC
+    HAS_KUBERNETES_COLLECTION = True
+except ImportError as e:
+    HAS_KUBERNETES_COLLECTION = False
+    k8s_collection_import_exception = e
+    K8S_COLLECTION_ERROR = traceback.format_exc()
+    K8sAnsibleMixin = object
+    AUTH_ARG_SPEC = {}
```
Comment (tadeboro) [inline plugins/modules/kubevirt_pvc.py]: No protection is needed.

File: plugins/modules/kubevirt_pvc.py [hunk 8]
```diff
@@ -305,9 +316,22 @@ class CreatePVCFailed(Exception):
...
+            supports_check_mode=True,
+        )
+        self.fail_json = self.module.fail_json
+
+        if not HAS_KUBERNETES_COLLECTION:
+            self.module.fail_json(
+                msg="The kubernetes.core collection must be installed",
+                exception=K8S_COLLECTION_ERROR,
+                error=to_native(k8s_collection_import_exception)
+            )
```
Comment (tadeboro) [inline plugins/modules/kubevirt_pvc.py]: Not needed.

File: plugins/modules/kubevirt_rs.py [hunk 9]
```diff
@@ -106,8 +106,14 @@
...
-
-from ansible_collections.community.kubernetes.plugins.module_utils.common import AUTH_ARG_SPEC
+try:
+    from ansible_collections.kubernetes.core.plugins.module_utils.args_common import AUTH_ARG_SPEC
+    HAS_KUBERNETES_COLLECTION = True
+except ImportError as e:
+    HAS_KUBERNETES_COLLECTION = False
+    k8s_collection_import_exception = e
+    K8S_COLLECTION_ERROR = traceback.format_exc()
+    AUTH_ARG_SPEC = {}
```
Comment (tadeboro) [inline plugins/modules/kubevirt_rs.py]: No protection is needed.

File: plugins/modules/kubevirt_template.py [hunk 10]
```diff
@@ -201,7 +201,15 @@
...
-from ansible_collections.community.kubernetes.plugins.module_utils.common import AUTH_ARG_SPEC
+try:
+    from ansible_collections.kubernetes.core.plugins.module_utils.args_common import AUTH_ARG_SPEC
+    from ansible_collections.kubernetes.core.plugins.module_utils.common import get_api_client
+    HAS_KUBERNETES_COLLECTION = True
+except ImportError as e:
+    HAS_KUBERNETES_COLLECTION = False
+    k8s_collection_import_exception = e
+    K8S_COLLECTION_ERROR = traceback.format_exc()
+    AUTH_ARG_SPEC = {}
```
Comment (tadeboro) [inline plugins/modules/kubevirt_template.py:212]: No protection is needed.

File: plugins/modules/kubevirt_vm.py [hunk 11]
```diff
@@ -245,7 +245,16 @@
...
-from ansible_collections.community.kubernetes.plugins.module_utils.common import AUTH_ARG_SPEC
+try:
+    from ansible_collections.kubernetes.core.plugins.module_utils.common import get_api_client
+    from ansible_collections.kubernetes.core.plugins.module_utils.args_common import AUTH_ARG_SPEC
+    HAS_KUBERNETES_COLLECTION = True
+except ImportError as e:
+    HAS_KUBERNETES_COLLECTION = False
+    k8s_collection_import_exception = e
+    K8S_COLLECTION_ERROR = traceback.format_exc()
+    AUTH_ARG_SPEC = {}
```
Comment (tadeboro) [inline plugins/modules/kubevirt_vm.py]: No protection is needed.

File: tests/unit/plugins/modules/kubevirt_fixtures.py [hunk 12]
```diff
@@ -2,15 +2,14 @@
 __metaclass__ = type
 
 import pytest
+import openshift.dynamic
```
Comment (tadeboro) [inline tests/unit/plugins/modules/kubevirt_fixtures.py]: This is a bit strange. I thought collection does not use openshift client anymore?

File: tests/unit/plugins/modules/kubevirt_fixtures.py [hunk 13]
```diff
@@ -66,8 +65,8 @@ def base_fixture(monkeypatch):
     openshift.dynamic.Resource.search = MagicMock()
     openshift.dynamic.Resource.watch = MagicMock()
```
Comment (tadeboro) [inline tests/unit/plugins/modules/kubevirt_fixtures.py]: Again, do we still need openshift client?

**Review by tadeboro — CHANGES_REQUESTED** [review]
I did the initial review, and the code looks OK in general. However, tests still reference the `openshift` Python library, which is strange since requirements do not list it anymore. So I guess this is something that still needs to be updated, right?

As for the failing tests on Ansible 2.9, I think the ansible-test's import machinery has troubles with the symlinks. I am not 100% sure, but since the `kubernetes.dynamic` is a symlink to the `kubernetes.base.dynamic`, this seems to be a reasonable explanation. Fixing this could be fun since Ansible 2.9 is in security-fixes-only maintenance mode.

Do note that the same tests also fail on `kubernetes.core`. Not sure why its CI did not pick this, though.

All in all, I think this PR is one significant step in the right direction. Thank you, @snecklifter, for your effort. It is greatly appreciated!

File: changelogs/fragments/k8s-core-2-compatibility.yml (hunk 1 above)
Comment (snecklifter) [reply changelogs/fragments/k8s-core-2-compatibility.yml]: thanks, have added

File: tests/unit/plugins/modules/kubevirt_fixtures.py (hunk 12 above)
Comment (snecklifter) [reply tests/unit/plugins/modules/kubevirt_fixtures.py]: @tadeboro its a good point, I hit lots of errors when moving to kubernetes.dynamic so thought I'd break down into two changesets but am re-attempting to debug the original errors now.

File: tests/unit/plugins/modules/kubevirt_fixtures.py (hunk 13 above)
Comment (snecklifter) [reply tests/unit/plugins/modules/kubevirt_fixtures.py]: See above, thanks for the review, much appreciated.

**Comment by jseguillon** [comment]
Hi everyone, 
As explained in #27 I'm especially concerned in keeping this collection alive since its a dependency for molecule KubeVirt driver I'm maintaining (https://github.com/jseguillon/molecule-kubevirt). 

@snecklifter are you still motivated in working on this issue ? 
@tadeboro same question for review ?

**Comment by tadeboro** [comment]
> @tadeboro same question for review?

I more or less stopped using Ansible and following the things happening in the community, so it would be better to find a reviewer with more up-to-date Ansible and community knowledge.

**Pull request closed without merging**